### ISSUE TITLE
Issue with WebGL Drawer and cached tiles when resizing the window

### ISSUE THREAD

**Issue by Tobio89** [openseadragon/openseadragon#2623]
I found an issue with the webgl drawer.
You can reproduce the issue in this codepen:
https://codepen.io/bruceji7/pen/mybJxPN

There isn't a lot of required setup to see the issue, but the steps needed to reproduce the issue are:
- Start with a viewer with webgl drawer
- Open the console to watch for errors
- Drag the window from one screen to another, smaller screen

I get this issue reliably if I drag the window from a monitor to my laptop's screen, since they are different sizes.

In the console it says `WebGL cannot be used to draw this TiledImage because it has tainted data. Does crossOriginPolicy need to be set?`
and then after that, there are lots of warnings such as this: `[Drawer._drawTileToCanvas] attempting to draw tile 10/1_0 when it's not cached`

It's clear that something causes the webgl drawer to believe it's tainted, and then to fall back to the canvas renderer. The same issue causes both the tainting and the warning about attempting to draw uncached tiles.

I know it's not crossOriginPolicy because that's set in the options, and if it's set wrong there are a lot of other errors that are very distinct.

**Comment by Tobio89** [comment]
In the Tile class, this line may be of interest:

```js
    getCanvasContext: function() {
        return this.context2D || (this.cacheImageRecord && this.cacheImageRecord.getRenderedContext());
    },
```

In my experiments, `this.context2D` has always been undefined even when drawing normally, but something seems to change in `cacheImageRecord` that causes a failure.

**Comment by pearcetm** [comment]
I'm unfortunately not able to reproduce this following the steps you've outlined using either Chrome or Firefox (Macbook Pro, M1 chip, OSX 13.3). Dragging a viewer from a larger external monitor back to my laptop monitor did not cause problems. What setup are you on?

**Comment by Tobio89** [comment]
I'm on an M1 Macbook pro too. I got this issue when dragging the window from a 1080p monitor to my macbook's screen.
It doesn't do it when I drag the screen between two monitors of the same size.

I'll try to get a screen recording of it.

**Comment by pearcetm** [comment]
> I'll try to get a screen recording of it.

I believe you that it happens, I just can't reproduce it locally on my setup which makes it hard to debug...

**Comment by Tobio89** [comment]
https://github.com/user-attachments/assets/632d95b4-28b4-4e63-b224-b56c88ca39ea

Originally I noticed this at the office, but I can reproduce it at home too.
Monitor resolutions are:
1920x1080 for the monitor, and 1728 x 1117 for the macbook's own display (weird size!)

**Comment by pearcetm** [comment]
You mentioned that something changes in `cacheImageRecord` - this is probably the place to look. @Aiosa any ideas? I know the cache system is changing soon but this is probably worth digging into anyway.

**Comment by Aiosa** [comment]
The issue comes from data not being cached properly for unknown reasons. It has nothign to do with tainted, the problem is that the drawer implementation considers data tainted even if they are missing (condition `if !data or $.isTainted(data)`). So we would like to probably rename the property to something more general (some flag 'I cannot draw this tiled image') + a cause message.

@Tobio89 can you debug [this method](https://github.com/openseadragon/openseadragon/blob/9b96707decbac98463c4deb43146f495e86f3b30/src/tilesource.js#L898)? it seems it fails to provide the canvas data. We are unable to reproduce the issue even with the demo, so I am afraid the debugging is up to you if you want to help us solve the issue.

**Comment by iangilman** [comment]
I can reproduce it. My guess is it's not about the resolution of the 2 monitors; it's the pixel density. I have a large HDPI screen (5k) and another normal density display (1000 x 1600). The image goes white if I drag from one monitor to the other. It doesn't matter which one I start on. I'm on Mac OS Sequoia (M1) and I tested on Chrome and Firefox.

**Comment by pearcetm** [comment]
@iangilman Glad you can reproduce, and the fact that pixel density is different seems like a promising lead. That seems like something that could be tested for and hopefully corrected, whether it just requires re-uploading the textures or maybe recreating the entire context. I'm not really sure to be honest.

**Comment by pearcetm** [comment]
With @iangilman's insight that it is the pixel density ratio that is the culprit here, this can be reproduced by zooming in/out within a browser tab. Changing pixel density ratio is already detected and triggers all existing tiles to be unloaded before redrawing the world:

https://github.com/openseadragon/openseadragon/blob/9b96707decbac98463c4deb43146f495e86f3b30/src/viewer.js#L2549-L2557

As was already noted, the issue isn't actually about tainted data, that error message was not specific. I also don't think the issue is actually with `WebGLDrawer` since I'm seeing issues with the canvas drawer too. It is rather about how the data needs to be reloaded and the drawer canvases updated after such a change.

**Comment by Aiosa** [comment]
I guess we could do better in overhaul than the current logics. I think invalidation pipeline trigger should be enough..?

**Comment by pearcetm** [comment]
> I think invalidation pipeline trigger should be enough..?

Resizing the actual drawing canvases to reflect the new pixel density ratio is needed too.

**Comment by Tobio89** [comment]
@Aiosa 
With a breakpoint on line909 `return cacheObject._renderedContext;`

This is during initial load, where it will behave correctly
![image](https://github.com/user-attachments/assets/dbf6bc03-5164-4c55-b3ac-daa9d1c11673)

After seeing the viewer go blank:
![image](https://github.com/user-attachments/assets/6efecc14-9db6-42b3-8c37-a85034bc6bac)

But I can't see any difference.

This breakpoint is called a lot, and I saw that immediately on drag, it worked fine, and then after stepping forward a bunch of times, it dropped.

**Comment by Aiosa** [comment]
Thanks for looking into it, meanwhile we found where lies the issue. I thought the cause is this method that provides the canvas data for tiles. Instead,  the reason of the bug is likely that there was detection of density ratio change, and the viewer threw away the data and called immediately draw(). It seems there is also a bug that, after calling reset, the tiled image keeps its state (e.g. target caches to draw). That's also a reason why the viewport goes blank, but moving (trigger subsequent draw) fixes it. @pearcetm submitted a PR that instead just calls forceResize, therefore we no longer throw away the data. But still, IMHO, the sequence of reset() and draw() is valid, and the root cause of the bug is still unsolved.

**Comment by pearcetm** [comment]
> the sequence of reset() and draw() is valid, and the root cause of the bug is still unsolved.

I agree @Aiosa, but since https://github.com/openseadragon/openseadragon/pull/2407 is so nearly complete I think we should look into this and fix it after that hits, if it isn't already fixed by that PR.

**Comment by iangilman** [comment]
Sounds good. We'll leave this issue open to investigate/validate after #2407.

**Comment by Aiosa** [comment]
I was actually unable to reproduce the issue (with page zoomming) on the overhaul PR. I am on the linked issue as well as on the project page.

**Comment by iangilman** [comment]
Sounds like this is fixed both by #2627 and #2407. I still think it's worth including #2627 so I've merged it. I think that closes this issue.

**Comment by Tobio89** [comment]
@iangilman 
Nice! Thanks for getting onto it so fast. Do you know when it might get released?
